### PR TITLE
Refactoring of the grunt build tool

### PR DESCRIPTION
**Pull request by taylortom**
#### What's new!
- All configs has been split into separate files per task! (see `grunt/config` - uses [load-grunt-config](https://github.com/firstandthird/load-grunt-config))
- All Tasks have been split into separate files! (see `grunt/tasks`)
- Everything is now loaded [asynchronously](https://github.com/adaptlearning/adapt_framework/blob/ABU-680/Gruntfile.js#L5-L14) as and when it's needed! (using [jit-grunt](https://www.npmjs.com/package/jit-grunt))
- New [custom Javascript compiler](https://github.com/adaptlearning/adapt_framework/blob/ABU-680/grunt/tasks/javascript.js)! (https://github.com/adaptlearning/adapt_framework/pull/931)
   - No more bundling! (components.js/extensions.js/menu.js/theme.js)
   - [Uses requirejs node module directly](https://github.com/adaptlearning/adapt_framework/blob/ABU-680/grunt/tasks/javascript.js#L51-L54)! (no more grunt-contrib-requirejs)
- New [custom LESS compiler](https://github.com/adaptlearning/adapt_framework/blob/ABU-680/grunt/tasks/less.js)! (with sourcemaps!)
- Tracking task code moved to local folder! (remove dependency on unsupported grunt plugin)
- Added ability to include/exclude specific plugins! (#723)
   * If **excludes** are specified, all plugins are included **excluding** those listed!
   * If **includes** are specified, **only** those plugins listed will be included!
   * If no includes/excludes are specified, all plugins are built!
- Build folder cleaned before a new build!
- Added code style checking/enforcement using jshint/jscs!
- Added 'time-grunt' to show execution time for tasks!

#### Instructions for testing!
1. Pull code from here!
1. Delete `node_modules` folder! 
1. Run `npm install && grunt build && grunt server`!
<br/>**If testing with the authoring tool, be aware that updating the git repo in `/temp` will cause certain deleted files/folders to be reinstated, so these will need to be manually removed (I think the main folder affected is `src/course`).**

### CODE REVIEW

**Comment by moloko** [comment]
There's something a bit wrong with the help for the `server-scorm` and `server` tasks:

Uploaded using [**ZenHub.io**](http://www.zenhub.io)

**Comment by taylortom** [comment]
Odd. Thanks Matt, will sort that

**Comment by moloko** [comment]
something's not quite right:
2016-01-18 19_39_31-cmder
I checked it wasn't anything in my setup (as I did switch over to using [nvm-windows](https://github.com/coreybutler/nvm-windows) the other days which has meant I've had to reinstall all my node packages) but I can still set up and build using `$ adapt create course` etc fine

**Comment by taylortom** [comment]
@moloko `path.parse` was new for node `v0.12`, so I'm assuming you're on `v0.10.33`? I've recently upgraded to the LTS, so haven't noticed this. Might have to work around using `parse`....

**Comment by moloko** [comment]
Yes that sounds likely - I was running v0.10.35

If I switch back to v4.2.4 it all works fine

**Comment by mediakitchen** [comment]
I gave this a test without any issues. I was testing on a Windows 10 machine. I am fairly new to this testing malarkey so essentially just downloaded the files from git and installed, built and previewed. Let me know if there are any specific things you want me to test.

File: grunt/tasks/javascript.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+		var _ = require('underscore');
+		var path = require("path");
+		var fs = require("fs");
+		var done = this.async();
+		var options = this.options({});
+
+		if (options.plugins) {
+			var pluginsClientSidePatch = 'requirejs.config({map: { "*": { "extensions/extensions":"'+options.pluginsModule+'","menu/menu":"'+options.pluginsModule+'","theme/theme":"'+options.pluginsModule+'","components/components":"'+options.pluginsModule+'" } } });';
+
+			if (!fs.existsSync(options.pluginsPath)) {
```
Comment (brian-learningpool) [inline grunt/tasks/javascript.js]: I think `existsSync()` has been deprecated.  Can you consider using `statSync()` here instead?

**Comment by oliverfoster** [comment]
i'm happy with the way this is working, so +1 for tom's bits

**Comment by taylortom** [comment]
@oliverfoster was just about to do that, thanks :grin:

**Comment by oliverfoster** [comment]
:hand: :palm_tree: :beers:

**Comment by oliverfoster** [comment]
can probably remove the console.log from https://github.com/adaptlearning/adapt_framework/blob/ABU-680/grunt/helpers.js#L51 now

**Comment by oliverfoster** [comment]
+1

**Comment by lc-thomasberger** [comment]
+1 (tested on windows)

**Comment by dancgray** [comment]
+1 (tested in Authoring tool on mac)

**Comment by moloko** [comment]
I'm assuming JSHint and JSCS still aren't fully integrated? I'm not seeing any alerts from them, is that as expected? If so then +1 (tested on Win 7 using Node v0.10.35 and [cmder](http://cmder.net/))

**Comment by brian-learningpool** [comment]
+1 -- tested on `develop` branch of Authoring Tool on OS X.  Great work @taylortom and @oliverfoster!

**Comment by taylortom** [comment]
@moloko correct, you need to specifically run the commands at the moment. I've created a [separate ticket](https://github.com/adaptlearning/adapt_framework/issues/933) to integrate these into travis.

If anyone else has some ideas about how we go about integrating this (if at all), will update this PR. @brian-learningpool @oliverfoster @dancgray @lc-thomasberger 

...otherwise, let's get this baby merged! :tada:

**Comment by oliverfoster** [comment]
go for it

**Pull request merged**